fix: event host unable to cancel event when `Disable Cancelling` is enabled.

GurneeshBudhiraja opened the pull request:
## What does this PR do?
The PR fixes the bug that was preventing the hosts to cancel their own events when `Disable Cancelling` is enabled.

- Fixes #24017 (GitHub issue number)
- Fixes CAL-6452 (Linear issue number - should be visible at the bottom of the GitHub issue description)

## Updates since last revision

Addressed review feedback and fixed additional issues:

1. **Fixed cancel action logic** (`bookingActions.ts`): The host bypass now also checks `isCancelled` and `isRejected` states to prevent hosts from attempting to cancel already cancelled/rejected bookings.

2. **Fixed missing `isHost` in dropdown** (`BookingActionsDropdown.tsx`): The `isHost` flag was computed but not included in the `actionContext` object, causing the host bypass to not work in the dropdown component. This was the root cause of the fix not working properly.

3. **Added test cases**: Two new tests verify that hosts cannot cancel already cancelled or rejected bookings.

## Visual Demo (For contributors especially)
#### Video Demo (if applicable):
The below video demo shows the updated functionality for both the host account(on the right) and the attendee account(on the left).
[Loom Video Link](https://www.loom.com/share/7a5b3a8066524c1f964afb1ff9c49d2a?sid=428c2909-7861-46b0-9123-ea2ed6e121cf)

## Mandatory Tasks (DO NOT REMOVE)

- [X] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [X] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox. N/A
- [X] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?
The hosts should be able to cancel the event even when `Disable Cancelling` is enabled. On the other hand, for somebody viewing the event other than the host, they would not be able to cancel it.

**Test scenarios:**
1. As a host with `Disable Cancelling` enabled, verify you CAN cancel an upcoming booking
2. As an attendee with `Disable Cancelling` enabled, verify you CANNOT cancel the booking
3. Verify that hosts cannot cancel already cancelled or rejected bookings (edge case)

## Checklist for human review

- [ ] Verify `isHost` is correctly computed in both `BookingListItem.tsx` (line 189-192) and `BookingActionsDropdown.tsx` (line 201)
- [ ] Verify the cancel action logic in `bookingActions.ts` (line 267) correctly handles the host bypass with cancelled/rejected checks

---
Link to Devin run: https://app.devin.ai/sessions/1f26c48a6898417188d7d856880e3f8b
Requested by: unknown ()

GurneeshBudhiraja committed:
Fixes #24017: Allow the host to cancel the event when `Disable Rescheduling` is enabled

GurneeshBudhiraja committed:
Update: Add the test in `bookingActions.test.ts` to test that the hosts can cancel event when cancellation is disabled

GurneeshBudhiraja committed:
Fixes #24017: Updates the logic for verifying the `isHost`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=24021) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=24021) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=24021) it.</sub>

GurneeshBudhiraja committed:
Merge branch 'main' into bug/host-unable-to-cancel-event

Devanshusharma2005 reviewed (commented):
Lets add loom to show the changes.

GurneeshBudhiraja commented on apps/web/components/booking/bookingActions.test.ts:
```diff
@@ -54,7 +54,7 @@ function createMockContext(overrides: Partial<BookingActionContext> = {}): Booki
           bookingId: 1,
           noShow: false,
-        } as any,
+        },
```
Adding `any` was giving the type issue at the time of commit.

GurneeshBudhiraja committed:
update `handleCancelBooking` utility to allow host cancellation when the event is disabled for cancellation

GurneeshBudhiraja committed:
Merge branch 'main' into bug/host-unable-to-cancel-event

KirankumarAmbati commented on apps/web/components/booking/BookingListItem.tsx:
```diff
@@ -189,6 +189,12 @@ function BookingListItem(booking: BookingItemProps) {
   const isAttendee = !!userSeat;
 
+  // Checks if the logged in user is the host of the booking
+  const isHost =
+    booking.user?.id != null &&
+    booking.loggedInUser.userId != null &&
+    booking.loggedInUser.userId === booking.user.id;
```
```suggestion
  const isHost = booking.loggedInUser?.userId === booking.user?.id;
```

pallava-joshi reviewed (changes requested):
looks good, can you please resolve the merge conflicts here.

GurneeshBudhiraja committed:
merge with main and resolve the merge conflicts

GurneeshBudhiraja committed:
prevent the host from cancelling the past meetings

GurneeshBudhiraja committed:
Changes: add more checks for the `userId` to determine the user host and update the test

GurneeshBudhiraja commented:
Hi @Pallava-Joshi would love if you could review this PR 😃

devin-ai-integration[bot] committed:
Merge upstream/main into bug/host-unable-to-cancel-event and resolve conflicts

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
Merge upstream/main into bug/host-unable-to-cancel-event and resolve conflicts

Co-Authored-By: unknown <>

romitg2 committed:
update webhook delete confirmation message

romitg2 committed:
revert

devin-ai-integration[bot] committed:
fix: prevent hosts from cancelling already cancelled/rejected bookings

Address review feedback from cubic-dev-ai:
- Add checks for isCancelled and isRejected in the isHost bypass
- Add tests to verify hosts cannot cancel already cancelled or rejected bookings

Co-authored-by: GurneeshBudhiraja <gurneeshbudhiraja@gmail.com>
Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: add missing isHost to BookingActionsDropdown actionContext

The isHost flag was computed but not included in the actionContext,
causing the host bypass for disableCancelling to not work in the
dropdown component.

Co-authored-by: GurneeshBudhiraja <gurneeshbudhiraja@gmail.com>
Co-Authored-By: unknown <>

romitg2 committed:
Merge branch 'main' into bug/host-unable-to-cancel-event

sahitya-chandra reviewed (commented):
Closing this in favor of #28051
